syntax = "proto3";

package ccfake;

// Approximation of CCF's KV interface
service KV
{
  rpc Ready (ReadyRequest) returns (BeginTx) {}

  // Get a single key
  rpc Get (GetRequest) returns (GetResponse) {}
  
  // Put a single key-value pair
  rpc Put (PutRequest) returns (PutResponse) {}

  rpc ApplyTx (ApplyRequest) returns (ApplyResponse) {}
}

message ReadyRequest {}

message BeginTx
{
  string uri = 1;
  bytes body = 2;
}

message GetRequest
{
  string table = 1;
  bytes key = 2;
}

message GetResponse
{
  optional bytes value = 1;
}

message PutRequest
{
  string table = 1;
  bytes key = 2;
  bytes value = 3;
}

message PutResponse
{
  bool existed = 1;
}

message ApplyRequest
{
  uint32 code = 1;
  bytes body = 2;
}

message ApplyResponse {}